src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -82,10 +75,12 @@ internal Tables(Node[] buckets, object[] locks, int[] countPerLock)
         /// </summary>
         private static bool IsValueWriteAtomic()
         {
-            Type valueType = typeof(TValue);
-
             // We don't have Type.IsClass, so we'll check if it's a reference type by seeing if it gets boxed;
             bool isClass = (object)default(TValue) == null;
+            if (isClass)
```

Comment:
`(object)default(decimal?) == null`, for example, is true but it's far from write-atomic.

Suggest:

```C#
if (default(decimal?) == null)
{
  return !typeof(TValue).IsConstructedGenericType || typeof(TValue).GetGenericTypeDefinition() != typeof(Nullable<>);
}